lmdk/modules/down_mixer/downmixer.c

```diff
@@ -0,0 +1,520 @@
+ * modules that need such processing should use proper wrappers
+ *
+ * process
+ *	- sources are handlers to source API struct source*[]
+ *	- sinks are handlers to sink API struct sink*[]
+ */
+static int downmix_process(struct processing_module *mod, struct sof_source **sources,
+			   int num_of_sources, struct sof_sink **sinks, int num_of_sinks)
+{
+	const struct module_self_data *const self = module_get_private_data(mod);
+
+	const uint8_t *input0_pos, *input0_start;
+	const uint8_t *input1_pos, *input1_start, *input1_end;
+	uint8_t *output_pos, *output_start;
+	unsigned int input1_channels;
+	size_t input1_frame_bytes;
+	size_t frames_processed, frames_to_process;
+	int32_t divider_input_0, divider_input_1;
+	size_t i;
+	int ret;
+
+	volatile int x = 0;
+	while (x);
+
+	frames_to_process = sink_get_free_frames(sinks[0]);
+	i = source_get_data_frames_available(sources[0]);
+	if (i < frames_to_process)
+		frames_to_process = i;
+
+	const unsigned int output_channels = sink_get_channels(sinks[0]);
+	const size_t output_frame_bytes = sink_get_frame_bytes(sinks[0]);
+	ret = sink_get_buffer(sinks[0], frames_to_process * output_frame_bytes, (void**)&output_pos,
+			      (void**)&output_start, &i);
+	if (ret)
+		return ADSP_FATAL_FAILURE;
+	const uint8_t * const output_end = output_start + i;
+
+	const unsigned int input0_channels = source_get_channels(sources[0]);
+	const size_t input0_frame_bytes = source_get_frame_bytes(sources[0]);
+	ret = source_get_data(sources[0], frames_to_process * input0_frame_bytes,
+			      (const void**)&input0_pos, (const void**)&input0_start, &i);
+	if (ret) {
+		sink_commit_buffer(sinks[0], 0);
+		return ADSP_FATAL_FAILURE;
+	}
+	
+	const uint8_t * const input0_end = input0_start + i;
+
+	/* If reference pin is not connected or module is in bypass mode, set input1_channels to 0.
+	 * This allows to skip reference pin content in the processing loop */
+	if (self->input1_channels_count_ && num_of_sources > 1 &&
+	    self->processing_mode_ != MODULE_PROCESSING_NORMAL) {
+		input1_channels = source_get_channels(sources[1]);
+		input1_frame_bytes = source_get_frame_bytes(sources[1]);
+		ret = source_get_data(sources[1], frames_to_process * input1_frame_bytes,
+				      (const void**)&input1_pos,
+				      (const void**)&input1_start, &i);
+		if (ret) {
+			sink_commit_buffer(sinks[0], 0);
+			source_release_data(sources[0], 0);
+			return ADSP_FATAL_FAILURE;
+		}
+		input1_end = input1_start + i;
+
+	} else {
+		input1_channels = 0;
+		input1_frame_bytes = 0;
+	}
+
+	/* Apply processing of the input chunks and generate the output chunk */
+	divider_input_0 = self->config_.divider_input_0;
+	divider_input_1 = self->config_.divider_input_1;
+
+	if (self->processing_mode_ == MODULE_PROCESSING_BYPASS) {
+		divider_input_0 = self->input0_channels_count_;
+		/* input1_channels is already set to 0 in BYPASS mode */
+	}
+
+	frames_processed = 0;
+	while (frames_processed < frames_to_process) {
```

Comment:
I'm struggling with that frames processing. For an outsider it is a bit complex and without knowing how pipelines work looks like magic. @marcinszkudlinski do we have any docs for that?